java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,943 @@
+import java
+private import RegexFlowConfigs
+
+/**
+ * A string literal that is used as a regular exprssion.
+ */
+abstract class RegexString extends Expr {
+  RegexString() { this instanceof StringLiteral }
+
+  /** Holds if a character set starts between `start` and `end`. */
+  private predicate char_set_start0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (
+      this.getChar(start + 1) = "^" and end = start + 2
+      or
+      not this.getChar(start + 1) = "^" and end = start + 1
+    )
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate char_set_end0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
+    not this.char_set_start0(_, pos - 1)
```

Comment:
With specifically `pos - 1` does this account for `[^]]`?

Reply:
Should do; since `char_set_start0` accounts for `[^`